File: learnLeetCode/Day47/Part3/platform_shifter.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <tuple>

using namespace std;

const long long INF = 1e18;

struct State {
    long long cost;
    int r, c;

    bool operator>(const State& other) const {
        return cost > other.cost;
    }
};

long long solvePlatformShifter(int n, int k, const vector<vector<int>>& grid) {
    priority_queue<State, vector<State>, greater<State>> pq;
    pq.push({0, 0, 0});

    vector<vector<long long>> dist(n, vector<long long>(n, INF));
    dist[0][0] = 0;

    int dr[] = {0, 0, 1, -1};
    int dc[] = {1, -1, 0, 0};

    while (!pq.empty()) {
        State current = pq.top();
        pq.pop();

        long long current_cost = current.cost;
        int r = current.r;
        int c = current.c;

        if (r == n - 1 && c == n - 1) {
            return current_cost;
        }

        if (current_cost > dist[r][c]) {
            continue;
        }

        for (int i = 0; i < 4; ++i) {
            int nr = r + dr[i];
            int nc = c + dc[i];

            if (nr >= 0 && nr < n && nc >= 0 && nc < n) {
                int h_curr = grid[r][c];
                int h_next = grid[nr][nc];

                long long move_cost = 1;
                long long penalty = 0;

                if (h_next > h_curr + k) {
                    penalty = h_next - (h_curr + k);
                }

                long long total_move_cost = move_cost + penalty;
                long long new_cost = current_cost + total_move_cost;

                if (new_cost < dist[nr][nc]) {
                    dist[nr][nc] = new_cost;
                    pq.push({new_cost, nr, nc});
                }
            }
        }
    }

    return -1;
}

int main() {
    int N = 3;
    int K = 2;
    vector<vector<int>> Grid = {
        {0, 5, 10},
        {2, 3, 8},
        {10, 1, 9}
    };

    long long result = solvePlatformShifter(N, K, Grid);
    cout << "Minimum Cost: " << result << endl;

    return 0;
}

File: Day47/Part3/platform_shifter.py
import heapq

def solve_platform_shifter(n, k, grid):
    # Dijkstra's Algorithm
    # Priority Queue stores: (cost, r, c)
    pq = [(0, 0, 0)]
    
    # Distances dictionary/grid initialized to infinity
    dist = {}
    dist[(0, 0)] = 0
    
    directions = [(0, 1), (0, -1), (1, 0), (-1, 0)]
    
    while pq:
        current_cost, r, c = heapq.heappop(pq)
        
        # If we reached the target
        if r == n - 1 and c == n - 1:
            return current_cost
        
        # Optimization: If we found a shorter path to this node already, skip
        if current_cost > dist.get((r, c), float('inf')):
            continue
        
        for dr, dc in directions:
            nr, nc = r + dr, c + dc
            
            if 0 <= nr < n and 0 <= nc < n:
                h_curr = grid[r][c]
                h_next = grid[nr][nc]
                
                move_cost = 1
                penalty = 0
                
                if h_next > h_curr + k:
                    penalty = h_next - (h_curr + k)
                
                total_move_cost = move_cost + penalty
                new_cost = current_cost + total_move_cost
                
                if new_cost < dist.get((nr, nc), float('inf')):
                    dist[(nr, nc)] = new_cost
                    heapq.heappush(pq, (new_cost, nr, nc))
                    
    return -1 # Should not happen if graph is connected

if __name__ == "__main__":
    N = 3
    K = 2
    Grid = [
        [0, 5, 10],
        [2, 3, 8],
        [10, 1, 9]
    ]
    
    result = solve_platform_shifter(N, K, Grid)
    print(f"Minimum Cost: {result}")
